OSDOCS-20042: adds OIDC authentication RHCL (#112868)
openshift/openshift-docs: modules/proc-rhcl-oidc-authpolicy.adoc

// Module included in the following assemblies:
//
// *deployment/deployment.adoc

:_mod-docs-content-type: PROCEDURE
[id="proc-set-oidc-auth-policy_{context}"]
= Setting the OIDC AuthPolicy custom resource

[role="_abstract"]

> You can configure a {prodname} `AuthPolicy` custom resource (CR) to use an OIDC provider. You can ensure that different clusters exposing the same API can authenticate with the same permissions.

Reviewer: ```suggestion
You can configure a {prodname} `AuthPolicy` custom resource (CR) to use an OIDC provider. The same API deployed to multiple clusters can authenticate with the same identity provider.
```

[TIP]
====
Apply an `AuthPolicy` custom resource (CR) with a `deny-all` policy to create a zero-trust environment. Using a zero-trust `AuthPolicy` object means that no traffic flows unless a specific allow rule is set. Every connection request must be authenticated. You can prevent the accidental exposing of services by using a `deny-all` policy.
====

In a zero-trust environment, every application with an `HTTPRoute` or `GRPCRoute` CR exposed by an application developer must also have an attached route-level `AuthPolicy` CR. You can attach multiple `AuthPolicy` objects to your `Gateway`, `HTTPRoute`, and `GRPCRoute` CRs.

.Prerequisites

* You installed {prodname} on one or more clusters.
* You installed the {oc-first}.
* You have write access to the {ocp} namespaces you need to work with.
* You have access to external or on-premise DNS.
* You created a `Gateway` object.

> * You created an `HTTPRoute` or `GRPCRoute` object.

Reviewer: ```suggestion
* You created an `HTTPRoute` or `GRPCRoute` object.
* You have a trusted OIDC provider configured to issue tokens for your API clients
```

.Procedure

. Create a default `AuthPolicy` CR, such as `gateway_name-auth.yaml`, with a `deny-all` setting for your `Gateway` object:
+
.Example AuthPolicy CR
[source,yaml]
----
apiVersion: kuadrant.io/v1
kind: AuthPolicy
metadata:
  name: ${KUADRANT_GATEWAY_NAME}-auth
  namespace: ${KUADRANT_GATEWAY_NS}
spec:
  authentication:
    "keycloak-kuadrant-realm":
      jwt:

>         issuerUrl: http://keycloak.keycloak.svc.cluster.local:8080/realms/kuadrant

Reviewer: If this is a deny-all policy, we probably do not want this authentication rule, but an `anonymous` one instead. The `authentication` field also goes under `rules`.

```suggestion
```

    response:
      unauthorized:
        headers:
          "content-type":
            value: application/json
        body:
          value: |
            {
              "error": "Forbidden",
              "message": "Access denied by default by the gateway operator. If you are the administrator of the service, create a specific auth policy for the route."
            }
----

. Apply your `AuthPolicy` CR by running the following command:
+
[source,terminal,subs="+quotes"]
----
$ oc apply -f _<gateway_name_auth.yaml>_
----
+
* Replace `_<gateway_name_auth.yaml>_` with the name of your `AuthPolicy` CR.

.Verification

. Check that your `AuthPolicy` has `Accepted` and `Enforced` status by running the following command:
+
[source,terminal]
----
$ oc get authpolicy ${KUADRANT_GATEWAY_NAME}-auth -n ${KUADRANT_GATEWAY_NS} -o=jsonpath='{.status.conditions[?(@.type=="Accepted")].message}{"\n"}{.status.conditions[?(@.type=="Enforced")].message}'
----
+
.Example output
[source,terminal]
----
AuthPolicy has been accepted
AuthPolicy has been successfully enforced
----

. Test your `AuthPolicy` CR by running the following `curl` command:
+
[source,terminal]
----

> $ curl -H 'Authorization: APIKEY ndyBzreUzF4zqDQsqSPMHkRhriEOtcRx' http://talker-api.127.0.0.1.nip.io:8000/hello

Reviewer: `APIKEY` is probably out of context.

I think we don't use `*.nip.io` hostnames for the downstream, do we?

----
+
.Example output
[source,text]
----

> HTTP/1.1 200 OK

Reviewer: If it's a default deny-all policy applied at the Gateway, without another one applied at the route, then the expected output is:

```suggestion
HTTP/1.1 403 Forbidden
```

In a zero-trust environment, every application with an `HTTPRoute` or `GRPCRoute` CR exposed by an application developer must also have an attached route-level `AuthPolicy` CR. You can target either a `Gateway` or a `route` object with an `AuthPolicy` CR.

.Prerequisites

* You installed {prodname} on one or more clusters.
* You installed the {oc-first}.
* You have write access to the {ocp} namespaces you need to work with.
* You have access to external or on-premise DNS.
* You created a `Gateway` object.
* You created an `HTTPRoute` or `GRPCRoute` object.
* You have a trusted OIDC provider configured to issue tokens for your API clients.
* If your OIDC server is using a self-signed or internal cluster certificate, you injected the trusted CA bundle into your {prodname} installation.

.Procedure

. Create an `AuthPolicy` CR that targets an `HTTPRoute` CR and configures OIDC authentication by using the following example:
+
.Example `AuthPolicy` CR
[source,yaml,subs="+quotes"]
----
apiVersion: kuadrant.io/v1
kind: AuthPolicy
metadata:

>   name: _<talker_api_oidc_auth>_

Reviewer: There's no explanation about the [Talker API](https://github.com/Kuadrant/authorino-examples/tree/main#talker-api) used in the example, leaving for the reader to figure out why this name was chosen for the AuthPolicy CR. WDYT about using a more generic name? E.g. `my-oidc-auth`.

  namespace: _<oidc_apps>_
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: HTTPRoute

>     name: talker-api-route

Reviewer: Similarly to https://github.com/openshift/openshift-docs/pull/112868/changes#r3473475738

```suggestion
    name: my-app-route
```

Author: Prod docs forbids "my", but this can be generic,  yes/

// Module included in the following assemblies:
//
// *deployment/rhcl-oids-authentication.adoc

:_mod-docs-content-type: PROCEDURE
[id="proc-set-oidc-auth-policy_{context}"]
= Setting the OIDC AuthPolicy custom resource

[role="_abstract"]
You can configure a {prodname} `AuthPolicy` custom resource (CR) to use an OIDC provider. The same API deployed to multiple clusters can authenticate with the same identity provider.

In a zero-trust environment, every application with an `HTTPRoute` or `GRPCRoute` CR exposed by an application developer must also have an attached route-level `AuthPolicy` CR. You can target either a `Gateway` or a `route` object with an `AuthPolicy` CR.

.Prerequisites

* You installed {prodname} on one or more clusters.
* You installed the {oc-first}.
* You have write access to the {ocp} namespaces you need to work with.

> * You have access to external or on-premise DNS.

Reviewer 2: I don't think that access to DNS is reqired for this guide.

* You created a `Gateway` object.
* You created an `HTTPRoute` or `GRPCRoute` object.
* You have a trusted OIDC provider configured to issue tokens for your API clients.
* If your OIDC server is using a self-signed or internal cluster certificate, you injected the trusted CA bundle into your {prodname} installation.

.Procedure

. Create an `AuthPolicy` CR that targets an `HTTPRoute` CR and configures OIDC authentication by using the following example:
+
.Example `AuthPolicy` CR
[source,yaml,subs="+quotes"]
----
apiVersion: kuadrant.io/v1
kind: AuthPolicy
metadata:
  name: _<user_apps>_

>   namespace: _<api_oidc_auth>_

Reviewer: ```suggestion
  name: _<api_oidc_auth>_
  namespace: _<user_apps>_
```

spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: HTTPRoute
    name: _<api_route>_
  rules:
    authentication:
      "oidc-idp-auth":
        jwt:
          issuerUrl: "https://rhbk.apps.cluster.example.com/realms/realm"
----
+
* The `AuthPolicy` CR must exist in the same namespace where the `HTTPRoute` object is applied.

> * Replace the `spec.metadata.name` parameter value with the name you want to use for your `AuthPolicy` CR.

Reviewer 3: ```suggestion
* Replace the `metadata.name` parameter value with the name you want to use for your `AuthPolicy` CR.
```

Author: This is actually wrong AFIAK. We have to use the entire reference.

> * Replace the `spec.metadata.namespace` parameter value with the namespace where you want your `AuthPolicy` CR applied.

Reviewer 3: ```suggestion
* Replace the `metadata.namespace` parameter value with the namespace where you want your `AuthPolicy` CR applied.
```
